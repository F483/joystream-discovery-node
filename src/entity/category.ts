import { Entity, Column, PrimaryColumn, BaseEntity } from "typeorm";

@Entity()
export class Category {

  @PrimaryColumn()
  id: number;

  @Column()
  parentId: number;

  @Column("text")
  title: string;

  @Column("text")
  description: string;

  @Column()
  createdAtBlockNumber: number;

  @Column()
  createdAtMoment: number;

  @Column()
  deleted: boolean;

  @Column()
  archived: boolean;

  @Column()
  numDirectSubcategories: number;

  @Column()
  numDirectModeratedThreads: number;

  @Column()
  positionInParentCategory: number;

  @Column()
  accountId: number;

  @Column()
  blockId: number;

  @Column()
  extrinsicIdx: number;

  @Column()
  eventIdx: number;
};