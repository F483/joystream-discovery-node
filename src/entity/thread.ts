import { Entity, Column, PrimaryColumn } from "typeorm";

@Entity("joystream_forum_thread")
export class Thread {

  @PrimaryColumn()
  id: number;

  @Column()
  title: string;

  @Column()
  text: string;

  @Column({
      name: 'category_id'
  })
  categoryId: number;

  @Column({
    name: 'nr_in_category'
  })
  nrInCategory: number;


  @Column({
    name: 'num_unmoderated_posts'
  })
  numUnmoderatedPosts: number;

  @Column({
    name: 'num_moderated_posts'
  })
  numModeratedPosts: number;

  @Column({
      name: 'created_at_block_number'
  })
  createdAtBlockNumber: number

  @Column({
    name: 'created_at_moment'
  })
  createdAtMoment: number

  @Column({
    name: 'author_id'
  })
  authorId: number;

  @Column({
    name: 'block_id'
  })
  blockId: number;

  @Column({
    name: 'extrinsic_idx'
  })
  extrinsicIdx: number;

  @Column({
    name: 'event_idx'
  })
  eventIdx: number;  

}