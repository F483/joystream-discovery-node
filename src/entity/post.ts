import {
  EntityManager,
  Like,
  Entity,
  Column,
  PrimaryColumn,
  ManyToOne,
  JoinColumn
} from "typeorm";
import { ForumThread } from "./thread";
import { ForumCategory } from "./category";
import { ForumModerationAction, ForumBlockchainTimestamp } from "./common";

/*
DESCRIBE joystream_forum_post; -- FIELD, TYPE, NULL, KEY, EXTRA
'id', 'bigint', 'NO', 'PRI', NULL, ''
'thread_id', 'bigint', 'YES', 'MUL', NULL, ''
'author_id', 'varchar(64)', 'YES', '', NULL, ''
'nr_in_thread', 'int', 'YES', '', NULL, ''
'current_text', 'text', 'YES', '', NULL, ''
'created_at_block_number', 'int', 'YES', '', NULL, ''
'created_at_moment', 'bigint', 'YES', '', NULL, ''
'block_id', 'int', 'NO', 'PRI', NULL, ''
'extrinsic_idx', 'int', 'YES', '', NULL, ''
'event_idx', 'int', 'YES', '', NULL, ''
*/
@Entity("joystream_forum_post")
export class ForumPost {
  public parent: ForumCategory;
  public replies: [ForumPost];
  public moderation: ForumModerationAction;
  // tslint:disable-next-line: variable-name
  public created_at: ForumBlockchainTimestamp;

  @PrimaryColumn()
  public id: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public thread_id: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public author_id: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public nr_in_thread: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public current_text: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public created_at_block_number: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public created_at_moment: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public block_id: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public extrinsic_idx: string;

  @Column()
  // tslint:disable-next-line: variable-name
  public event_idx: string;

  /*
  @ManyToOne(
    type => ForumThread,
    thread => thread.posts
  )
  @JoinColumn({ name: "thread_id" })
  public thread: ForumThread;
  */
}

export async function findPost(
  manager: EntityManager,
  text: string
): Promise<ForumPost[]> {
  return manager.find(ForumPost, {
    current_text: Like(`%${text}%`)
  });
}

/*
  type ForumPost {
    # Relations
    thread: ForumThread
    # author: MembersProfile # TODO add when harvester updated

    # From: apps/packages/joy-types/src/forum.ts:PostType
    id: string # instead of u64 due to overflow
    thread_id: String # instead of u64 due to overflow
    nr_in_thread: string # instead of u32 due to overflow
    current_text: string
    moderation: ForumModerationAction
    created_at: ForumBlockchainTimestamp
    author_id: string # AccountId
  }
*/
