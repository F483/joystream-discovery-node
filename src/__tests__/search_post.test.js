import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import sinon from 'sinon';

import { constructTestServer } from './__utils';
import { EntityManager } from 'typeorm';
import { PostDB } from '../../dist/src/entity/post';

const SEARCH_POSTS = gql`
    query postList($term: String) {
        getPosts(text: $term) {
            id
            current_text
            thread {
                id
                title
                text
            }
        }
    }
`;

describe('Search posts query', () => {

    it('finds a match for existing string', async () => {
        const manager = sinon.createStubInstance(EntityManager);
        let post = new PostDB()
        post.id = 1
        post.title = 'post 1'
        post.text = 'post 1 text'
        manager.find.resolves([post]);

        const context = {
            manager: manager
        }
        const { server } = constructTestServer({context: context});
        const {query} = createTestClient(server);
        const res = await query({query: SEARCH_POSTS, variables: {term: 'test term'}});
        expect(res).toMatchSnapshot();
    });
});
