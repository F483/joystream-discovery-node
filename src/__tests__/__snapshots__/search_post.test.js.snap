// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search posts query finds a match for existing string 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Cannot query field "current_text" on type "Post". Did you mean "currentText"?],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
